File: backend/GameApp.Domain/ValueObjects/Scenes/ItemScene.cs
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Items;

namespace GameApp.Domain.ValueObjects.Scenes;

// Class for a scene where you get an item
public class ItemScene : Scene
{

    public Item RewardItem { get; private set; } = default!;


    // Constructor
    public ItemScene(SceneName name, SceneDescription description, Biomes biome, Item item)
    : base(name, description, biome)
    {
        RewardItem = item;
    }

    // Getter
    public Item GetRewardItem() => RewardItem;

    // Setters 
    public ItemScene SetSceneName(SceneName newName) =>
    new ItemScene(newName, GetDescription(), GetBiome(), RewardItem);

    public ItemScene SetSceneDescription(SceneDescription newDescription) =>
    new ItemScene(GetName(), newDescription, GetBiome(), RewardItem);

    public ItemScene SetBiome(Biomes newBiome) =>
    new ItemScene(GetName(), GetDescription(), newBiome, RewardItem);

    public ItemScene SetItem(Item newItem) =>
    new ItemScene(GetName(), GetDescription(), GetBiome(), newItem);

    // To string
    public override string ToString()
    {
        return $"{GetName()} Item Scene: " +
               $"Description={GetDescription()}, Biome={GetBiome()}, " +
               $"Reward Item={RewardItem}";
    }
}

File: frontend/gameapp-frontend-js/src/domain/entities/scenes/item-scene.js
import { Item } from "../items/item.js";
import { Scene } from "./scene.js";

// ItemScene
export class ItemScene extends Scene {
    constructor(name, description, biome, item, id = null) {
        super(name, description, biome, id);
        if (!(item instanceof Item)) throw new TypeError("item must be Item");
        this._rewardItem = item;
    }

    // getter
    get rewardItem() { return this._rewardItem; }

    // setter
    setSceneName(newName) { return new ItemScene(newName, this._description, this._biome, this._rewardItem, this._id); }
    setSceneDescription(newDescription) { return new ItemScene(this._name, newDescription, this._biome, this._rewardItem, this._id); }
    setBiome(newBiome) { return new ItemScene(this._name, this._description, newBiome, this._rewardItem, this._id); }
    setItem(newItem) { return new ItemScene(this._name, this._description, this._biome, newItem, this._id); }

    toString() {
        return `${this._name} Item Scene(${this._id}): Description=${this._description}, Biome=${this._biome}, Reward Item=${this._rewardItem}`;
    }
}
